lib/spack/spack/test/cmd/pkg.py

```diff
@@ -236,3 +236,51 @@ def test_pkg_fails_when_not_git_repo(monkeypatch):
     monkeypatch.setattr(spack.cmd, 'spack_is_git_repo', lambda: False)
     with pytest.raises(spack.main.SpackCommandError):
         pkg('added')
+
+
+def test_pkg_source(mock_packages):
+    fake_source = pkg("source", "fake")
+
+    fake_file = spack.repo.path.filename_for_package_name("fake")
+    with open(fake_file) as f:
+        contents = f.read()
+        assert fake_source == contents
+
+
+def test_pkg_canonical_source(mock_packages):
+    source = pkg("source", "multimethod")
+    assert "@when('@2.0')" in source
+    assert "Check that multimethods work with boolean values" in source
+
+    canonical_1 = pkg("source", "--canonical", "multimethod@1.0")
+    assert "@when" not in canonical_1
+    assert "should_not_be_reached by diamond inheritance test" not in canonical_1
+    assert "return 'base@1.0'" in canonical_1
+    assert "return 'base@2.0'" not in canonical_1
+    assert "return 'first_parent'" not in canonical_1
+    assert "'should_not_be_reached by diamond inheritance test'" not in canonical_1
+
+    canonical_2 = pkg("source", "--canonical", "multimethod@2.0")
+    assert "@when" not in canonical_2
+    assert "return 'base@1.0'" not in canonical_2
+    assert "return 'base@2.0'" in canonical_2
+    assert "return 'first_parent'" in canonical_2
+    assert "'should_not_be_reached by diamond inheritance test'" not in canonical_2
+
+    canonical_3 = pkg("source", "--canonical", "multimethod@3.0")
+    assert "@when" not in canonical_3
+    assert "return 'base@1.0'" not in canonical_3
+    assert "return 'base@2.0'" not in canonical_3
+    assert "return 'first_parent'" not in canonical_3
+    assert "'should_not_be_reached by diamond inheritance test'" not in canonical_3
+
+    canonical_4 = pkg("source", "--canonical", "multimethod@4.0")
+    assert "@when" not in canonical_4
+    assert "return 'base@1.0'" not in canonical_4
+    assert "return 'base@2.0'" not in canonical_4
+    assert "return 'first_parent'" not in canonical_4
+    assert "'should_not_be_reached by diamond inheritance test'" in canonical_4
+
+
+def test_pkg_hash(mock_packages):
+    pkg("hash", "multimethod")
```

Comment:
Is this just testing that the `spack pkg hash multimethod` command doesn't fail? Could we test against its actual hash now that the hash is consistent?

Reply:
The hash is only consistent if the package doesn't change.  So testing that would require another argument to grab the source from some alternative and unchanging place, but we already have other tests that do that... I didn't think it was worth the trouble.

Reply:
Ok, great!